### DISCUSSION TITLE
DVC Plugin to skip Nodes if Data and Code are up to date

### DISCUSSION THREAD

**Discussion by Spooky-0** [kedro-org/kedro#837] [Idea]
Hi All,

I have been thinking of creating a Kedro Plugin which leverages DVCs Version Controlled Data but Kedros beautiful pipelines. Somewhere I have read a discussion of this idea before, but I can’t find it anymore. Maybe Github, Discord, etc.

Background: At work I had the problem of jobs taking a long time and failing often, but once they ran successfully, all was good. However, when changing parts of the pipeline, after a while, you don’t know if your data is up to date any more - so you have to rerun all. And this takes a long time. So I used DVC to track my input data (e.g. CSV, Hadoop Data*, etc.) but I also tracked the scripts (.py, .sh, etc.). When any one of those changed, the Pipeline would run only that what was necessary. For development this was a game changer.
*) I added an extra step, which simply called ‘DESCRIBE my_hive/impala_table’, saved the output somewhere, and then DVC’ed just this proxy. If the Data on Hive changed, DVC would not know - but this was fine.

For Kedro I was thinking of creating a Plugin/Hook which has very similar behaviour. A Nodes Input is very easily tracked by DVC (especially easy for local data). The Code is not so easily tracked.
Some difficulties: I cannot track all the code, because if I change Node17 then I would have to rerun everything, because accoarding to DVC some code changed.
Thus I want to have a DVC run only look at the current Nodes Code and Data. I was thinking of using python’s inspect module to track the code for this Node. I think this would work. The problem might be, that the Node imports code from a different file. If this file then changes, would DVC know? Is it possible to track all relevant code?

DVC can only run one task at a time. Thus integrating with kedro-accelrator would be impossible. Right?

To reiterate: The goal of kedro-dvc would not be the Data Version Control per se, but rather it would help during devlopment of kedro pipelines. You could be sure that the code/data is always up to date, but only have to run individual nodes - the rest can be skipped as the data is up to date.
The goal is not to keep the data versioned (I guess you should use versioned datasets xD)

Before I embark on this, I would like to ask the community on feedback.

Thanks!

**Comment by mzjp2** [comment]
Relevant discussions/resources:

* https://github.com/quantumblacklabs/kedro/issues/528
* https://discourse.kedro.community/t/speeding-up-pipeline-processing-with-change-detection/90/9
* https://gist.github.com/mzjp2/076bfd73b0215bda01ee71186966389d

I hadn't though of using DVC, but on the face of it, that sounds like a awesome way to implement the functionality.

Funnily enough, my gist above got a comment by @bvancil (https://github.com/bvancil) that also asked about the code-versioning functionality. I thought along the same lines as you, that comparing the AST of the node would be a cool way to do this, but your point on module-level things like imports changing is something I hadn't considered. Potentially it could be node-code + module-code (minus other functions/nodes)?

**Reply by datajoely** [reply]
Hi @Spooky-0  - this is something the core team are looking at building out (medium term) under the category of 'change data capture' via the `KedroSession`. In the short term I'd also like to point you to [kedro-dolt](https://www.dolthub.com/blog/2021-06-16-kedro-dolt-plugin/) plugin which can do some very clever things indeed.

**Reply by Spooky-0** [reply]
Hi @datajoely , is there already code/discussion available for 'change data capture'? I couldn't find any on github.
thanks!

**Reply by Spooky-0** [reply]
Thanks @mzjp2 , thats exactly the conversation i was refering to. Especially the part about using `inspect` to validate if a node changed.
My ideas would probably be: Iteration 1) Use input code file and input data file. If any changed, rerun the node.
Iteration 2) Use only relevant code from code file.
Iteration 3) Use all relevant code -> imported functions,modules, etc.

DVC would need a cache to save the relevant code. So, i guess the kedro project structure would get an additional
`.dvc` folder and `dvc.yaml` for standard dvc stuff. But also `.kedro-dvc` folder, which would contain all types of code snippets to compare. E.g. `.kedro-dvc/nodes/mypipe/node_name.py` which has all code aggregated.

The dvc run would then look like this: `dvc run -d .kedro-dvc/nodes/mypipe/node_name.py -d data/01_raw/test.csv -o data/02_.../clean.csv -c "kedro run --target node_name"` (or similar, I forget the exact notation) This is how I did it at my job.
But its not very nice, because now dvc forks and creates a child process to run `kedro run ...`. Thus I would like to use the python api of dvc to do this stuff inside of my hook.

I have played around with it a bit yesterday, but ran into a bit of an Issue.
It seems that Node Hooks are run in parallel, as opposed to in serries (pipeline). See my amazing ascii image:  

Expected/Wished pipeline behaviour: `input->[hook1]->[hook2]->[node]->ouput`  

Actual parallel behaviour 
```
input -> [hook1]
input -> [hook2]
input -> [node] -> output
```  


Current Relevant Code:
```
# https://github.com/quantumblacklabs/kedro/blob/c2ca44ece1c57b51df4cd2cbff6aee8f8f75088c/kedro/runner/runner.py#L233
def _collect_inputs_from_hook(
    node: Node,
    catalog: DataCatalog,
    inputs: Dict[str, Any],
    is_async: bool,
    run_id: str = None,
) -> Dict[str, Any]:
    inputs = inputs.copy()  # shallow copy to prevent in-place modification by the hook
    hook_manager = get_hook_manager()
    hook_response = hook_manager.hook.before_node_run(  # pylint: disable=no-member
        node=node,
        catalog=catalog,
        inputs=inputs,
        is_async=is_async,
        run_id=run_id,
    )

    additional_inputs = {}
    for response in hook_response:
        if response is not None and not isinstance(response, dict):
            response_type = type(response).__name__
            raise TypeError(
                f"`before_node_run` must return either None or a dictionary mapping "
                f"dataset names to updated values, got `{response_type}` instead."
            )
        response = response or {}
        additional_inputs.update(response)

    return additional_inputs
```
being called by :
```
#  https://github.com/quantumblacklabs/kedro/blob/c2ca44ece1c57b51df4cd2cbff6aee8f8f75088c/kedro/runner/runner.py#L302
    additional_inputs = _collect_inputs_from_hook(
        node, catalog, inputs, is_async, run_id=run_id
    )
    inputs.update(additional_inputs)
```
I want my hooks to potentially change the node function, (to be just `pass` or sth). Any ideas?

**Reply by meaningfromdata** [reply]
@datajoely Any updates on this skipping of nodes if data has not changed and "change data capture versioning" in kedro?  I have a similar use case where I only want to store new versions if the raw data inputs have changed.  At this point is the versioning in kedro just done by timestamp without any inspection of the data inputs?

**Reply by datajoely** [reply]
For now its still timestamp based - we have a team experimenting with DVC, will update on the results

**Reply by shaunc** [reply]
Excellent -- I'd be a happy user of kedro-dvc integration. Was researching how to do it myself....

**Reply by shaunc** [reply]
@Spooky-0 -- instead of using inspect, perhaps one could leverage a "designated test generator" for a given artifact (type?) and [coverage](https://coverage.readthedocs.io/en/6.3/config.html?highlight=output#xml), parsing the xml or json output.

*Afterthought* -- thinking about our own code, the problem for us with this approach would be numba.cuda code, which isn't covered. *Usually* we place cuda compiled code together with wrapper code in the same module, so as long as we weren't trying to resolve beyond the module level this would work. I presume that perhaps inspect could do a bit better (seeing the call, at least) but it would also have problems with calls out into c code, etc. Presumably some config/manual specification of relation between certain code and certain artifacts could close this gap.

... perhaps pipeline steps can be associated with tests (and perhaps many-many?) via a wrapper. Could use [coverage contexts](https://coverage.readthedocs.io/en/stable/contexts.html#dynamic-contexts) to mark what code was run by which of these tests (in CI/CD, for instance).

**Reply by antonymilne** [reply]
@shaunc LOVE the idea of using something like coverage to do this. This sounds much more promising to me than trying to parse the AST ourselves. Why try and reinvent the wheel to work out which lines of Python are executed when other people have already figured that out? Also might be worth considering [trace](https://docs.python.org/3/library/trace.html) and [heartrate](https://github.com/alexmojaki/heartrate).

Being compatible with CUDA would be nice, but honestly I think that's quite an unusual/advanced case you have. If we could have something working just with normal Python code then that would be amazing. Even just the naive `inspect` of the node's function or looking at all modules within a node's directory would be a start.

**Reply by shaunc** [reply]
Heartrate seems higher level, but trace looks quite nice... hmm... so one question would be to do this "at runtime" -- as you could with trace, or via a test as I was thinking. 

My idea was that you either declared a test with the node, or you wrapped a test in a decorator that referenced the node. Vs CUDA or etc etc my idea was in that wrapper you could also specify some modules to track manually.

I guess you could do this using trace automagically without having a designated test? I guess there is a performance hit vs not having to maintain the test and seeing what branches actually ran in a node execution.

**Reply by antonymilne** [reply]
Yeah, I was imagining something where you would do this at runtime rather than needing to write a test for each node that you want to "version control". Otherwise it seems like a bit of an abuse of the test runner, which we're only really using as a vehicle to get coverage. And you'd have to ensure that you're running the tests in between every `kedro run`, which maybe couples you to doing it in CI/CD only? Depending on the library used, definitely it might affect performance if we're doing this run time though.

The other possible issue I see with the test runner strategy is that you need to write test data for every single node you want to version control. I was thinking that you could avoid writing an individual test for each node by just writing one parametrised test that does something clever by going through each of kedro's nodes. But you'd still need a way to run each node function on its own test input data somehow.

**Reply by shaunc** [reply]
Yes -- I guess you could perhaps have the best of both worlds if there was a "record mode" where you would run and write per-node-metadata on source code usage (which you could check in), and a "production mode" where you would just use this data to possibly skip nodes or not without having trace on.

I personally would still like the ability to add additional code manually to the generated list (and not have it clobbered by rerecord, of course).

**Comment by shaunc** [comment]
The flip-side of this discussion on partial rebuilds -- suppose I'm using argo-workflows to run my pipeline. I want to get all the artifacts into DVC so that I can later use them for partial rebuilds. Should I write wrappers that check in outputs? What about logs and error outputs? Also in a distributed environment I can't just rely on the local work tree -- it would seem that I need a versioning policy to guide how the commits are merged.

Has any of this work been done already? Suggestions on where to look, or where to put it if I want to build something myself?

*Update* -- [dvc checkpoints](https://dvc.org/doc/user-guide/experiment-management/checkpoints) seem to be the thing to integrate with. I guess the approach would be to make kedro the source of authority, and automate generation of dvc.yaml, etc....  (unless there is some reason to think about doing it the other way around?)

**Reply by shaunc** [reply]
@Spooky-0, @AntonyMilneQB @mzjp2 -- is there any code out there that generates dvc.yaml-s from pipeline definitions? Is that part of the vision here? I would be happy to work on this -- should it go in a kedro-dvc repo/plugin?

**Reply by datajoely** [reply]
So I'm not sure we have anything off the shelf - since joining the LF we've had a lot of activity on the topic so we'll try and consolidate it all shortly (see this discssion for details https://github.com/kedro-org/kedro/discussions/1181) 

Also happy to answer any questions about the internals that may be helpful - there will be some small breaking changes in 0.18.0 that we can keep front of mind.

**Reply by shaunc** [reply]
Thanks, @datajoely ... on the dvc side I've gotten some encoragement --
https://discord.com/channels/485586884165107732/485596304961962003/938108282311147552
https://discord.com/channels/485586884165107732/485596304961962003/938119057109037108
Once I consolidate and think through things I'll spin up a `kedro-dvc` repo with a design document (to start out with, at least :)) -- probably next week. Comments and contributions will be welcome!

[*Update* -- [more info on the dvc side](https://discord.com/channels/485586884165107732/938821298929430548/939175277228072970)]

**Comment by shaunc** [comment]
An initial stab at a design -- https://github.com/FactFiber/kedro-dvc/blob/main/doc/design.md
Comments welcome!

BTW ... 
```
The current project's Python requirement (>=3.8,<3.10) is not compatible with some of the required packages Python requirement:
    - kedro requires Python >=3.6, <3.9, so it will not be satisfied for Python >=3.9,<3.10
 ``` 

:( ...
Are you adding 3.9 support soon? (If numba can do it, so can you! :)) -- I like to try to use numpy typing which worked (works?) better with 3.9 (I think that was the issue....). I could go lower in this package, but the package I want to consume this in definitely needs 3.9.

**Reply by datajoely** [reply]
3.9 (and 3.10) support will be out shortly as part of 0.18.0 which should come out by the end of Q1 2022.

Thank you for pushing that - it looks awesome. We recently had a conversation with some other open source users who highlighted the sheer cost difference of writing DVC deltas vs the entire dataset on each run. I for one am really excited about this and also agree with @AntonyMilneQB that the `coverage` idea is genius!

**Reply by antonymilne** [reply]
This looks super interesting! I'm going to have to take some time to learn more about DVC to make any intelligent comments on it but will definitely do so when I'm more wised up 😀 

On the 3.9 support: this is coming with kedro 0.18 but if you're just trying to develop a plugin then I think you should probably be able to use it anyway - see https://github.com/kedro-org/kedro/discussions/1117. Note that kedro 0.18 will still support Python 3.7 and 3.8 though, so depending on how essential the improved numpy typing support is we should think about the benefit of taking advantage of those features vs. the cost of incompatibility for 3.7/3.8 users.

**Reply by datajoely** [reply]
Also feel free to tag us in PRs and issues on your repo if it's to do with Kedro internals and we'll do our best ot help :)

**Reply by shaunc** [reply]
We need 3.9 support certainly, but we can try to keep 3.8 at the very least as well for the moment. Hmm... how to handle with CI? Are you perchance going to create an 0.18 pre-release (which is stable enough! :)) anytime soon? I encourage it! :)

As for comments -- I'm sure there is still much to think through. The first issue that comes to mind is versioning. I haven't looked into it yet, but would hypothesize that dvc uses git versioning tools to produce hashes. Perhaps this is a good way to go? Is there a point at which I can intercept how you generate version strings and substitute in the dvc version hash? I encourage you to take the discussion to the [plugin discussion board](https://github.com/FactFiber/kedro-dvc/discussions) :)

Thank you!

**Reply by antonymilne** [reply]
0.18 pre-release is unofficially available already in the `develop` branch. `main` always points to the latest development on the minor release branch (currently 0.17.7), while `develop` points to the latest development on the major release branch (currently 0.18.0). So if you want to work off 0.18 then you can do so already by `pip install https://github.com/kedro-org/kedro/archive/develop.zip`. Definitely not officially supported, but it should actually be pretty stable I think. 

Will take a look at the plugin discussion board for the other stuff! Very exciting 😀

**Reply by shaunc** [reply]
develop -- almost!:
```shell
 % poetry add https://github.com/kedro-org/kedro/archive/develop.zip
...
  Because no versions of black match >22.1.0,<23.0.0
   and black (22.1.0) depends on click (>=8.0.0), black (>=22.1.0,<23.0.0) requires click (>=8.0.0).
  And because kedro (0.17.6) depends on click (<8.0), black (>=22.1.0,<23.0.0) is incompatible with kedro (0.17.6).
```
sigh -- I wish poetry would allow overrides....  however -- works with black `21.12b0`! :) ... still -- if you can loosen that click dependency -- probably I won't be the only one....

**Reply by antonymilne** [reply]
Ah yes, [this was pointed out last week](https://github.com/kedro-org/kedro/issues/1193) and is on the backlog to fix 🙂

**Comment by shaunc** [comment]
I've written some more detail into the [basic correspondence between kedro and dvc](https://github.com/FactFiber/kedro-dvc/discussions/6).

A question -- is there a mechanism for attaching extra metadata to data catalog entries, nodes and pipelines? Perhaps a generic mechanism would suffice. E.g. a meta=dict(kedro_dvc=dict( ...)) or equiv in yaml for data catalog.

This could also be provided externally, but it gets harder to be DRY that way.

* catalog entries: cached or uncached? Other dvc metadata
* nodes: e.g. whether to turn into checkpoints in kedro experiments
* pipelines: naming experiments, mapping persistent experiments onto directories or tags...

**Reply by datajoely** [reply]
Hi @shaunc  thanks for pushing things here!

So the question on generic metadata has come up before and it's something I've personally been interested in because of the potential of generating data docs. This is being tracked in this open issue https://github.com/kedro-org/kedro/issues/1076.

In terms of short term solution - there is potentially a route forward with a generic 'wrapper dataset' which works the same way as our [PartitionedDataSet ](https://kedro.readthedocs.io/en/stable/kedro.io.PartitionedDataSet.html) that essentially passes the configuration to another dataset also provided. You could maybe define a `KedroMetaDataSet` or something that allows for aribtary or `dvc` specific metadata?

You could also consider is injecting run parameters provided by the `--params` syntax or possibly custom CLI arguments that are then used to manage the 'experiment' mapping.

Lastly, if you were to utilse Kedro's beta [experiment tracking features](https://kedro.readthedocs.io/en/latest/08_logging/02_experiment_tracking.html) as a way of identifying nodes which include 'tracked' information?

**Reply by shaunc** [reply]
@datajoely -- thanks for the reply!

What I've settled on for the moment is to put additional metadata in `conf/kedro_dvc` -- with subdirectories `data/` for data-related, and `pipeline/` for node and pipeline related. (See [data discussion](https://github.com/FactFiber/kedro-dvc/discussions/9) for example.)

These directories will also include generated `.dvc` / `dvc.yaml` files (respectively), and these will themselves have extra metadata space. I will do my best to leave non-generated fields untouched, so that if you want to do dvc-specific things (external repo info, etc), it will be preserved.

Eventually the plan is to add some utilities to help maintain synchronization (`kedro dvc data mv` ...). At the same time we could add "porcelain" wrappers to make this process more seamless. (maybe variants of `DVCDataset`, also `kd_node`, `kd_pipeline`). If you guys move forward on metadata (hopefully for nodes and pipelines as well as datasets), these wrappers might be unnecessary, or at least would be very lightweight.

As for the experiment tracking features. One mismatch is that DVC supports plots and metrics as experiment outcomes, your features support metrics, while plots are included together with datasets. But I think this is fine -- I can use your experiment tracking datasets to identify what outputs are metrics, while using plot dataset types to identify plots.

In a sense, your experiment tracking specifies what is necessary -- a json file could be parameters, a data source or metrics output. But it would be nice if perhaps the API moved in the direction of specifying what is a diagnostic and what is an intermediate output orthogonally. (Maybe have "Metric" /"Plot" mixins that simply mark outputs. These in turn could simply set metadata on the underlying datasets, which could themselves concentrate on IO provision....)

Also note that I have decided, at first at least, **not** to support kedro versioned datasets. The rationale is that, once you have crossed the threshold to git integration, rather than use working-tree "space" for versions, it is better to use the git "time" dimension (branches, commits, or more ephemeral [dvc exp references](https://dvc.org/doc/user-guide/experiment-management/experiments-overview#dvc-experiments-overview)):
> Experiments are custom [Git references](https://dvc.org/blog/experiment-refs) (found in .git/refs/exps) with one or more commits based on HEAD. These commits are hidden and not checked out by DVC. Note that these are not pushed to Git remotes by default either (see [dvc exp push](https://dvc.org/doc/command-reference/exp/push)).

Using space for experiments as a whole (ie pipeline runs) may sometimes be a good idea. DVC supports publishing experiments to git branches, but [suggests that sometimes space-wise organization is the way to go](https://dvc.org/doc/user-guide/experiment-management#organization-patterns).

I would envision using kedro modules to keep track of experiment versions that you wanted to keep track of in the repo simultaneously. Perhaps I could add some utility to facilitate publishing an experiment to a worktree location.  (NB perhaps one easy way to do this is via [git-worktree](https://git-scm.com/docs/git-worktree).)

**Reply by shaunc** [reply]
@datajoely -- the documentation for "cli hooks" is ... sort of missing. :) Are there examples you could point me to if I wanted to add arguments to `kedro run`, and then use them in a `before_pipeline_run` hook?

For example, if I wanted to by default skip nodes that dvc says don't need recreation, but wanted to add a `--force` (or `--force-recreate`) option to ignore that? Or should I be thinking in terms of not effecting how the standard commands work, and provide parallel functionality (e.g. `kedro dvc repro ...`, `kedro dvc exp run ...`)?

**Reply by antonymilne** [reply]
Just to be clear, there's a few different concepts here:
1. [CLI hooks](https://kedro.readthedocs.io/en/latest/07_extend_kedro/02_hooks.html#cli-hooks) (docs just added, only available in `latest`), which enables you to inject behaviour using `before_command_run`. Not sure whether it will be useful here 
2. Injecting custom parameters into `kedro run` using `--params`. This is already possible without any further customisation. Anything you enter here will become `run_params` dictionary in the `before_pipeline_run` hook. It's quite a limited syntax currently though so you can't enter "complex" values like dictionaries - basically `kedro run --params key:1value1,key2:value2`. If you want to make `run_params` (or part of it) available to other hooks that don't have `run_params` explicitly available to them, you can save it as an instance attribute - check [this Discord message](https://discord.com/channels/778216384475693066/846330075535769601/938376592479096862) onwards.
3. If you need to add a whole new argument to the `run` command, this is also possible. The docs for this do indeed seem to be mysteriously missing, but I found some of them on the `develop` branch... https://github.com/kedro-org/kedro/blob/develop/docs/source/extend_kedro/common_use_cases.md#use-case-3-how-to-add-or-modify-cli-commands. Basically it's the same as overwriting any `kedro xyz` command. I wonder if there's a nicer way of doing this than just copying and pasting the existing kedro `def run` actually - let me have a think about that.

In terms of how to structure the CLI (extra options vs. parallel functionality), this is a good question and I don't think there's an obvious "right" way of doing it - really it depends on what functionality the plugin ends up providing. I would encourage you though to take a look at the plugins-integrations channel on Discord which is frequented by other plugin authors (and users) who might have more thoughts on it. e.g. I'd guess that @Galileo-Galilei has thought about this before with the kedro-mlflow plugin.

The DVC experiment concept looks very interesting but I'm afraid I'm going to have to take some more time to understand it fully to be able to comment intelligently on how kedro should interact with it...

**Reply by datajoely** [reply]
Hi @shaunc off topic for a second - it would be great to join our [discord server](https://discord.gg/akJDeVaxnB) - it may make sense to make a dedicated `kedro-dvc` channel, I'd also like to personally send you some Kedro swag as a thank you so DM me on discord when you get a chance!

**Reply by shaunc** [reply]
@datajoely -- thanks for the invite! @AntonyMilneQB -- thanks for the replies!

* Thanks for the links to additional documentation.
* Both kedro and dvc allow extra command line args. Kedro-dvc should support, but at first I'm focusing on node-declared parameters as I have to support those as a minimum.
* When I wrote the message 2 days ago, it seemed to me that the best was, if kedro-dvc was installed, to "take over" the standard commands so as to minimize the surface area for users. However, what if you want to run some pipelines "dvc-aware" and others not? Also kedro-dvc will at first only support a subset of kedro features. So it seems best to me now to provide separate commands (with "dvc" command group) at first. Perhaps there will be an install recipe to configure full integration.

**Reply by shaunc** [reply]
@datajoely -- do you have an "after_command_run" cli hook? We wanted to add our own conf install after `kedro new` runs if `kedro-dvc` is installed?

**Reply by antonymilne** [reply]
@shaunc that doesn't exist, but we could probably add it if it's useful. The `before_command_run` hook was added for a specific purpose (telemetry) and there was no need for `after_command_run` so it wasn't added. The closest thing at the moment would be `after_pipeline_run`, but obviously that's only invoked if you're running a pipeline.

**Reply by shaunc** [reply]
It would be useful... its not essential though. Right now I'm starting with a sequence `kedro new` then `kedro dvc install` with the latter creating the DVC repo and K-D config on the inside. But if you have K-D installed globally, though it would be convenient to do these steps in `kedro new` (perhaps modulated by an option).

**Comment by dhavala** [comment]
Hello Kedro Devs/Mods,

Any updates in this work?

I read that Kedro provides data versioning -- but it simply writes data to folder with timestamp. 
While that is a good starting point -- but taking a more principled and solid approach is needed, such as those taken by DVC (which seperates the metadata from data, and only tracks the metadata via git).

Recently, there are some demo's that show how MLFlow and DVC can work together. I am wondering are there any efforts to make DVC work with Kedro Data Catalogue.

It solves many problems and 
core Kedro funct. is modular, opiniated pipelines. avoids working on duplicated responsibiliies
mlflow for tracking model, serving etc
dvc for data versioning
cml for CI/CD 

Then Kedro can play nicely along with other mature open source tools in the MLOps space. Otherwise, it feels as if, everybody (tool) becomes a siwss-army knife of MLops.

Appreciate any pointers or comments or suggestions!

**Reply by datajoely** [reply]
@shaunc how are things going? Is there any support we cna provide?
